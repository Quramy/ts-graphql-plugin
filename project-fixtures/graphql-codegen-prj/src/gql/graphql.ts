/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
};

export type Post = {
  __typename?: 'Post';
  author: User;
  body: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  title: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  popularPosts: Array<Post>;
};

export type User = {
  __typename?: 'User';
  avatarURL: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type PopularPosts_QueryQueryVariables = Exact<{ [key: string]: never; }>;


export type PopularPosts_QueryQuery = { __typename?: 'Query', popularPosts: Array<(
    { __typename?: 'Post', id: string }
    & { ' $fragmentRefs'?: { 'PostSummary_PostFragment': PostSummary_PostFragment } }
  )> };

export type PostSummary_PostFragment = { __typename?: 'Post', id: string, title: string, author: (
    { __typename?: 'User', name: string }
    & { ' $fragmentRefs'?: { 'UserAvatar_UserFragment': UserAvatar_UserFragment } }
  ) } & { ' $fragmentName'?: 'PostSummary_PostFragment' };

export type UserAvatar_UserFragment = { __typename?: 'User', name: string, avatarURL: string } & { ' $fragmentName'?: 'UserAvatar_UserFragment' };

export const UserAvatar_UserFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<UserAvatar_UserFragment, unknown>;
export const PostSummary_PostFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<PostSummary_PostFragment, unknown>;
export const PopularPosts_QueryDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"PopularPosts_Query"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"popularPosts"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"PostSummary_Post"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}}]} as unknown as DocumentNode<PopularPosts_QueryQuery, PopularPosts_QueryQueryVariables>;