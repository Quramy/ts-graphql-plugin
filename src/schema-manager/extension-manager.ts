import path from 'path';
import { GraphQLSchema, parse, extendSchema, DocumentNode, GraphQLError } from 'graphql';
import { SchemaBuildErrorInfo } from './schema-manager';
import { SchemaManagerHost } from './types';

export class ExtensionManager {
  private _targetSdlFileNames: string[];
  private _parsedExtensionAstMap = new Map<string, { node: DocumentNode; sdlContent: string }>();
  private _graphqlErrorMap = new Map<string, SchemaBuildErrorInfo>();

  constructor(private _host: SchemaManagerHost) {
    const { localSchemaExtensions } = this._host.getConfig();
    this._targetSdlFileNames = (localSchemaExtensions || []).map(filePath =>
      this._getAbsoluteSchemaPath(this._host.getProjectRootPath(), filePath),
    );
  }

  readExtensions() {
    this._targetSdlFileNames.forEach(filePath => this._readExtension(filePath));
  }

  extendSchema(baseSchema: GraphQLSchema) {
    if (this._graphqlErrorMap.size) return null;
    for (const [fileName, { node, sdlContent }] of this._parsedExtensionAstMap.entries()) {
      try {
        baseSchema = extendSchema(baseSchema, node);
      } catch (error) {
        if (error instanceof Error) {
          const { message } = error;
          this._graphqlErrorMap.set(fileName, { message, fileName, fileContent: sdlContent });
        }
        return null;
      }
    }
    this._graphqlErrorMap.clear();
    return baseSchema;
  }

  getSchemaErrors() {
    return [...this._graphqlErrorMap.values()];
  }

  startWatch(cb: () => void, interval = 100) {
    this._targetSdlFileNames.forEach(fileName => {
      this._host.watchFile(
        fileName,
        () => {
          this._host.log('Changed local extension schema: ' + fileName);
          this._readExtension(fileName);
          cb();
        },
        interval,
      );
    });
  }

  private _readExtension(fileName: string) {
    if (!this._host.fileExists(fileName)) return null;
    const sdlContent = this._host.readFile(fileName, 'utf8');
    if (!sdlContent) return null;
    this._host.log('Read local extension schema: ' + fileName);
    try {
      const node = parse(sdlContent);
      this._parsedExtensionAstMap.set(fileName, { node, sdlContent });
      this._graphqlErrorMap.delete(fileName);
    } catch (error) {
      if (error instanceof GraphQLError) {
        const { message, locations } = error;
        this._host.log('Failed to parse: ' + fileName + ', ' + message);
        if (locations) {
          this._graphqlErrorMap.set(fileName, {
            message,
            fileName,
            fileContent: sdlContent,
            locations: locations.map(loc => ({ line: loc.line - 1, character: loc.column - 1 })),
          });
        } else {
          this._graphqlErrorMap.set(fileName, { message, fileName, fileContent: sdlContent });
        }
      }
    }
  }

  private _getAbsoluteSchemaPath(projectRootPath: string, schemaPath: string) {
    if (path.isAbsolute(schemaPath)) return schemaPath;
    return path.resolve(projectRootPath, schemaPath);
  }
}
