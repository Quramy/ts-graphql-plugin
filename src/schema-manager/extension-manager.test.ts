import fs from 'fs';
import path from 'path';
import { buildSchema, printSchema } from 'graphql';
import { ExtensionManager } from './extension-manager';
import { createTestingSchemaManagerHost } from './testing/testing-schema-manager-host';

function createManagerWithHost(config: { localSchemaExtensions: string[] }) {
  const host = createTestingSchemaManagerHost({
    schema: '',
    localSchemaExtensions: config.localSchemaExtensions,
    files: config.localSchemaExtensions.map(name => ({
      fileName: path.join(__dirname, name),
      content: fs.readFileSync(path.join(__dirname, name), 'utf-8'),
    })),
    prjRootPath: __dirname,
  });
  return { extensionManager: new ExtensionManager(host), host };
}

function createManager(config: { localSchemaExtensions: string[] }) {
  return createManagerWithHost(config).extensionManager;
}

const baseSdl = `
  type Query {
    hello: String!
  }
`;

const baseSchema = buildSchema(baseSdl);

describe(ExtensionManager, () => {
  it('should parse and extend base schema', () => {
    const extensionManager = createManager({ localSchemaExtensions: ['./testing/resources/normal.graphql'] });
    extensionManager.readExtensions();
    const schema = extensionManager.extendSchema(baseSchema);
    expect(printSchema(schema!)).toMatchSnapshot();
    expect(extensionManager.getSchemaErrors()).toStrictEqual([]);
  });

  it('should store parser errors with invalid syntax file', () => {
    const extensionManager = createManager({ localSchemaExtensions: ['./testing/resources/invalid_syntax.graphql'] });
    extensionManager.readExtensions();
    const schema = extensionManager.extendSchema(baseSchema);
    expect(schema).toBeNull();
    const errors = extensionManager
      .getSchemaErrors()!
      .map(e => ({ ...e, fileName: e.fileName!.replace(__dirname, '') }));
    expect(errors).toMatchSnapshot();
  });

  it('should store parser errors with invalid extension', () => {
    const extensionManager = createManager({
      localSchemaExtensions: ['./testing/resources/invalid_extension.graphql'],
    });
    extensionManager.readExtensions();
    const schema = extensionManager.extendSchema(baseSchema);
    expect(schema).toBeNull();
    const errors = extensionManager
      .getSchemaErrors()!
      .map(e => ({ ...e, fileName: e.fileName!.replace(__dirname, '') }));
    expect(errors).toMatchSnapshot();
  });

  it('should execute call back when files change', async () => {
    const { extensionManager, host } = createManagerWithHost({
      localSchemaExtensions: ['./testing/resources/normal.graphql'],
    });
    const called = new Promise<void>(res => extensionManager.startWatch(res));
    host.updateFile(path.join(__dirname, 'testing/resources/normal.graphql'), '');
    await called;
  });
});
