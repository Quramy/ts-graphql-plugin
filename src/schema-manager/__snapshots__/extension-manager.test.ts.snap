// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtensionManager should parse and extend base schema 1`] = `
"type Query {
  hello: String!
  me: User!
}

type User {
  name: String!
}
"
`;

exports[`ExtensionManager should store parser errors with invalid extension 1`] = `
Array [
  Object {
    "fileContent": "extend type NoExisitingType {
  name: String!
}
",
    "fileName": "/testing/resources/invalid_extension.graphql",
    "message": "Cannot extend type \\"NoExisitingType\\" because it is not defined.",
  },
]
`;

exports[`ExtensionManager should store parser errors with invalid syntax file 1`] = `
Array [
  Object {
    "fileContent": "directive @hoge() on FIEELD
",
    "fileName": "/testing/resources/invalid_syntax.graphql",
    "locations": Array [
      Object {
        "character": 16,
        "line": 0,
      },
    ],
    "message": "Syntax Error: Expected Name, found \\")\\".",
  },
]
`;
